File: utils/rateLimiter.js
// Simple in-memory rate limiter
// For production use, consider Redis or a database-backed solution

const rateLimits = new Map();

// Rate limit configurations (command: { maxUses, windowMs })
const RATE_LIMIT_CONFIG = {
    logsolo: { maxUses: 10, windowMs: 60000 }, // 10 uses per minute
    myrefines: { maxUses: 5, windowMs: 30000 }, // 5 uses per 30 seconds
    leaderboard: { maxUses: 3, windowMs: 60000 }, // 3 uses per minute
    setrate: { maxUses: 2, windowMs: 300000 }, // 2 uses per 5 minutes
    resetstats: { maxUses: 1, windowMs: 600000 } // 1 use per 10 minutes
};

/**
 * Check if a user is within rate limits for a command
 * @param {string} userId - The user ID to check
 * @param {string} command - The command being executed
 * @returns {boolean} - True if within rate limits, false if exceeded
 */
function checkRateLimit(userId, command) {
    const config = RATE_LIMIT_CONFIG[command];
    
    if (!config) {
        // No rate limit configured for this command
        return true;
    }

    const key = `${userId}:${command}`;
    const now = Date.now();
    
    // Get or create user's rate limit data
    if (!rateLimits.has(key)) {
        rateLimits.set(key, {
            count: 0,
            resetTime: now + config.windowMs
        });
    }

    const userLimit = rateLimits.get(key);

    // Check if the window has expired
    if (now >= userLimit.resetTime) {
        userLimit.count = 0;
        userLimit.resetTime = now + config.windowMs;
    }

    // Check if user has exceeded the limit
    if (userLimit.count >= config.maxUses) {
        return false;
    }

    // Increment the counter
    userLimit.count++;
    return true;
}

/**
 * Get remaining uses for a user and command
 * @param {string} userId - The user ID to check
 * @param {string} command - The command to check
 * @returns {object} - Object with remaining uses and reset time
 */
function getRemainingUses(userId, command) {
    const config = RATE_LIMIT_CONFIG[command];
    
    if (!config) {
        return { remaining: Infinity, resetTime: null };
    }

    const key = `${userId}:${command}`;
    const now = Date.now();
    
    if (!rateLimits.has(key)) {
        return { remaining: config.maxUses, resetTime: null };
    }

    const userLimit = rateLimits.get(key);

    // Check if the window has expired
    if (now >= userLimit.resetTime) {
        return { remaining: config.maxUses, resetTime: null };
    }

    return {
        remaining: Math.max(0, config.maxUses - userLimit.count),
        resetTime: userLimit.resetTime
    };
}

/**
 * Reset rate limits for a specific user and command
 * @param {string} userId - The user ID to reset
 * @param {string} command - The command to reset (optional, resets all if not provided)
 */
function resetUserRateLimit(userId, command = null) {
    if (command) {
        const key = `${userId}:${command}`;
        rateLimits.delete(key);
    } else {
        // Reset all commands for the user
        for (const key of rateLimits.keys()) {
            if (key.startsWith(`${userId}:`)) {
                rateLimits.delete(key);
            }
        }
    }
}

/**
 * Clear expired rate limit entries (cleanup function)
 */
function cleanupExpiredLimits() {
    const now = Date.now();
    
    for (const [key, limit] of rateLimits.entries()) {
        if (now >= limit.resetTime) {
            rateLimits.delete(key);
        }
    }
}

/**
 * Get rate limit configuration for a command
 * @param {string} command - The command to get config for
 * @returns {object|null} - Rate limit configuration or null if not found
 */
function getRateLimitConfig(command) {
    return RATE_LIMIT_CONFIG[command] || null;
}

// Clean up expired entries every 5 minutes
setInterval(cleanupExpiredLimits, 300000);

module.exports = {
    checkRateLimit,
    getRemainingUses,
    resetUserRateLimit,
    cleanupExpiredLimits,
    getRateLimitConfig
};

File: utils/rate_limiter.py
import time
from typing import Dict, Tuple

class RateLimiter:
    def __init__(self):
        # Store rate limit data: {user_id: {command: (count, reset_time)}}
        self.rate_limits: Dict[str, Dict[str, Tuple[int, float]]] = {}
        
        # Rate limit configurations (command: (max_uses, window_seconds))
        self.rate_limit_config = {
            'logsolo': (10, 60),      # 10 uses per minute
            'myrefines': (5, 30),     # 5 uses per 30 seconds
            'leaderboard': (3, 60),   # 3 uses per minute
            'spicesplit': (3, 300),   # 3 uses per 5 minutes
            'help': (5, 60),          # 5 uses per minute
            'setrate': (2, 300),      # 2 uses per 5 minutes
            'resetstats': (1, 600)    # 1 use per 10 minutes
        }

    def check_rate_limit(self, user_id: str, command: str) -> bool:
        """Check if a user is within rate limits for a command"""
        config = self.rate_limit_config.get(command)
        if not config:
            # No rate limit configured for this command
            return True

        max_uses, window_seconds = config
        now = time.time()
        
        # Initialize user data if not exists
        if user_id not in self.rate_limits:
            self.rate_limits[user_id] = {}
        
        if command not in self.rate_limits[user_id]:
            self.rate_limits[user_id][command] = (0, now + window_seconds)

        count, reset_time = self.rate_limits[user_id][command]

        # Check if the window has expired
        if now >= reset_time:
            count = 0
            reset_time = now + window_seconds

        # Check if user has exceeded the limit
        if count >= max_uses:
            return False

        # Increment the counter
        count += 1
        self.rate_limits[user_id][command] = (count, reset_time)
        return True

    def get_remaining_uses(self, user_id: str, command: str) -> Tuple[int, float | None]:
        """Get remaining uses for a user and command"""
        config = self.rate_limit_config.get(command)
        if not config:
            return 999999, None

        max_uses, window_seconds = config
        now = time.time()
        
        if (user_id not in self.rate_limits or 
            command not in self.rate_limits[user_id]):
            return max_uses, None

        count, reset_time = self.rate_limits[user_id][command]

        # Check if the window has expired
        if now >= reset_time:
            return max_uses, None

        return max(0, max_uses - count), reset_time

    def reset_user_rate_limit(self, user_id: str, command: str | None = None):
        """Reset rate limits for a specific user and command"""
        if user_id in self.rate_limits:
            if command:
                self.rate_limits[user_id].pop(command, None)
            else:
                # Reset all commands for the user
                self.rate_limits[user_id] = {}

    def cleanup_expired_limits(self):
        """Clean up expired rate limit entries"""
        now = time.time()
        users_to_remove = []
        
        for user_id, user_limits in self.rate_limits.items():
            commands_to_remove = []
            
            for command, (count, reset_time) in user_limits.items():
                if now >= reset_time:
                    commands_to_remove.append(command)
            
            for command in commands_to_remove:
                del user_limits[command]
            
            if not user_limits:
                users_to_remove.append(user_id)
        
        for user_id in users_to_remove:
            del self.rate_limits[user_id]